src/config.rs

```diff
@@ -3787,4 +3788,82 @@ mod tests {
             }
         );
     }
+
+    #[test]
+    fn test_config_template_is_valid() {
+        let template_config = std::include_str!("../etc/config-template.toml")
+            .lines()
+            .map(|l| l.strip_prefix('#').unwrap_or(l))
+            .join("\n");
+
+        let mut cfg: TiKvConfig = toml::from_str(&template_config).unwrap();
+        cfg.validate().unwrap();
+    }
+
+    #[test]
+    fn test_config_template_no_superfluous_keys() {
+        let template_config = std::include_str!("../etc/config-template.toml")
+            .lines()
+            .map(|l| l.strip_prefix('#').unwrap_or(l))
+            .join("\n");
+
+        let mut deserializer = toml::Deserializer::new(&template_config);
+        let mut unrecognized_keys = Vec::new();
+        let _: TiKvConfig = serde_ignored::deserialize(&mut deserializer, |key| {
+            unrecognized_keys.push(key.to_string())
+        })
+        .unwrap();
+
+        // Don't use `is_empty()` so we see which keys are superfluous on failure.
+        assert_eq!(unrecognized_keys, Vec::<String>::new());
+    }
+
+    #[test]
+    fn test_config_template_matches_default() {
+        let template_config = std::include_str!("../etc/config-template.toml")
+            .lines()
+            .map(|l| l.strip_prefix('#').unwrap_or(l))
+            .join("\n");
+
+        let mut cfg: TiKvConfig = toml::from_str(&template_config).unwrap();
+        let mut default_cfg = TiKvConfig::default();
+
+        // Some default values are computed based on the environment.
+        // Because we can't set config values for these in `config-template.toml`, we will handle
+        // them manually.
+        cfg.readpool.unified.max_thread_count = default_cfg.readpool.unified.max_thread_count;
```

Comment:
What is the value in `cfg` if not set in example, why does it differ from the default?

Reply:
You are right, if they are not set in `config-template.toml`, they don't differ from the default.
When I wrote that code looked through *all* `Default` implementations on config items and added special rules for all computed ones, even if they are not set explicitly in `config-template.toml`. The idea was to avoid failures when somebody sets them in `config-template.toml` in the future.
If you want, we can also remove them because they don't make any difference.

Reply:
> The idea was to avoid failures when somebody sets them in config-template.toml in the future

I think this is intended: to fail if the template is not relflecting the default value.

Reply:
hm okay I see.

Just to wrap things up:
 * yes, the test should fail if a config differs from the `Default` implementation.
 * some configs are computed. thus, a fixed value in `config-template.toml` might be equal to the `Default` value on some machines, but not on others. This would lead to test failures on some machines. So we just overwrite these computed values.
 * if a value is missing in `config-template.toml`, it falls back to the `Default` value.

In this case, `max_thread_count` is set to `8` in `config-template.toml`, but actually it's a computed value that differs from machine to machine. The idea was to get rid of spurious test failures by just "ignoring" the value from `config-template.toml` and overwriting it with `Default`. The downside of this approach is that `max_thread_count` could be set to _anything_ in the config file.

But you are right: some of the computed values are **not** set in `config-template.toml` and they will then be set to `Default` while parsing. Thus, we wouldn't need to overwrite them in the test. However, if someone decides to add them to `config-template.toml` because he or she wants to document it, it might introduce test failures on some machines because it's a computed value.
So all in all I would actually keep these "overwrites" for **all** computed values in the test. But also I might lack the experience of judging this well, so please feel free to correct me and we can find a different solution :)

Reply:
Oh! it sounds reasonable, or else we will not be able to list computed config in the example.